### DISCUSSION TITLE
IMAP observations

### DISCUSSION THREAD

**Discussion by cbrake** [emersion/maddy#463] [General]
This is not a complaint (Maddy is generally working well, and I know IMAP support is beta) -- simply an observation -- occasionally, when I read a message in Thunderbird, the message does not get marked as read. If I navigate away from the message and come back, then it usually gets marked as read. Again, not a huge deal, but thought I would start collecting observations here and possibly help debug this. I don't recall this being an issue with my old postfix/dovecot setup.

If this is observed by others, then perhaps it makes sense to move to an issue.

**Comment by cbrake** [comment]
Another observation today -- sent a message and got an error in thunderbird that the save failed. Tried again, and it succeeded. Saw the following in the logs -- not sure if its related.

```
Mar 08 20:38:36 mail5 maddy[42586]: imapsql: CreateMessage (uidNext) []: database is locked         {"mbox":"Sent","mboxId":20,"username":"cbrake@<some domain>","uid":4}
```

I occasionally see the `database is locked` messages at other times. Using sqlite.

**Reply by vtolstov** [reply]
Yes i have similar problems with sqlite backend sometimes. I think that this is concurrent issue when two writers try to lock database

**Comment by cbrake** [comment]
would changing the DB backend improve problems with messages not getting marked read or deleted? It seems Thunderbird has a really hard time staying in sync over IMAP.

**Comment by smlx** [comment]
This is a maddy bug, and addressed in the [`go-sqlite3` FAQ](https://github.com/mattn/go-sqlite3#faq) (see the last item).

**Reply by foxcpp** [reply]
It is fundamental sqlite restriction. maddy already uses `cache=shared` among other options to make it less of a problem. However, the most complete solution wound be to automatically retry such operations until they succeed without returning any errors to the client.

**Reply by smlx** [reply]
Does maddy also set `db.SetMaxOpenConns(1)`?

**Reply by foxcpp** [reply]
No, it does not as that would significantly degrade performance.

**Comment by cbrake** [comment]
Would the switching the DB to Postgres reduce the IMAP sync problems I'm seeing, or is that a separate issue?

**Comment by cbrake** [comment]
Another thing that several of us have noticed is that a significant number of messages don't get deleted out of the drafts folder after an email is sent. Setup:

* Thunderbird client
* Maddy with SQLite database
